File: blusage.cpp
#include "blusage.h"

#include <QDomDocument>
#include <tidy/tidy.h>
#include <tidy/buffio.h>

BLUsage::BLUsage(){
    QDate today = QDate::currentDate();
    start = QDate(today.year(), today.month(), 1);
    end = start.addDays(29);
}

QString BLUsage::errorString() {
    return error;
}

bool BLUsage::parse(QString html) {
    // tidy the horrible html
    html = tidy(qPrintable(html));

    if (html.isEmpty()) {
        error = "Malformed HTML. Could not tidy it.";
        return false;
    }
    // remove all unwanted/invalid elements
    html = html.toLower();
    html.replace("&nbsp;", "");

    QList<QStringList> elems;
    elems.append(QStringList() << "<head" << "</head>");
    elems.append(QStringList() << "<script" << "</script>");
    elems.append(QStringList() << "<!--" << "-->");

    int left, right;
    QStringList l;
    foreach (l, elems) {
        forever {
            left = html.indexOf(l[0]);
            right = html.indexOf(l[1]);
            if (left == -1)
                break;
            html = html.left(left) + html.right(html.size() - right - l[1].size());
        }
    }

    left = html.indexOf("<body");
    right = html.indexOf(">", left);
    html = html.left(left + 5) + html.right(html.size() - right);

    QDomDocument doc("blusage");
    if (!doc.setContent(html)) {
        error = "Could not find data.";
        return false;
    }

    QDomNodeList rows = doc.documentElement().elementsByTagName("table").at(14).childNodes();
    if (rows.isEmpty()) {
        error = "Could not find data.";
        return false;
    }

    usage.clear();
    QString date;
    QList<QStringList> timely;
    int dataKB;
    totalKB = 0;
    for (int i=1; i<rows.size()-1; i++) {
        QDomNodeList tds = rows.at(i).childNodes();
        QStringList l;
        for (int j=2; j<5; j++) {
            l << tds.at(j).firstChildElement().text();
        }
        if (date != l[0]) {
            if (!date.isEmpty()) {
                DailyUsage daily;
                daily.day = date;
                daily.dataUsed = dataKB;
                daily.detail = timely;
                totalKB += dataKB;
                usage.append(daily);
            }
            timely.clear();
            date = l[0];
            dataKB = 0;
        }
        dataKB += l[2].toInt();
        timely.append(QStringList() << l[1] << l[2]);
    }

    lastUpdate = QDateTime::currentDateTime();

    return true;
}

QString BLUsage::tidy(const char* input) {
    TidyBuffer output = {0};
    TidyBuffer errbuf = {0};
    int rc = -1;
    Bool ok;

    TidyDoc tdoc = tidyCreate();                     // Initialize "document"

    ok = tidyOptSetBool( tdoc, TidyXhtmlOut, yes );  // Convert to XHTML
    if ( ok )
      rc = tidySetErrorBuffer( tdoc, &errbuf );      // Capture diagnostics
    if ( rc >= 0 )
      rc = tidyParseString( tdoc, input );           // Parse the input
    if ( rc >= 0 )
      rc = tidyCleanAndRepair( tdoc );               // Tidy it up!
    if ( rc >= 0 )
      rc = tidyRunDiagnostics( tdoc );               // Kvetch
    if ( rc > 1 )                                    // If error, force output.
      rc = ( tidyOptSetBool(tdoc, TidyForceOutput, yes) ? rc : -1 );
    if ( rc >= 0 )
      rc = tidySaveBuffer( tdoc, &output );          // Pretty Print

    QString html = (rc >= 0) ? QString((const char*)output.bp) : "";

    tidyBufFree( &output );
    tidyBufFree( &errbuf );
    tidyRelease( tdoc );

    return html;
}

QDataStream &operator<<(QDataStream &out, const DailyUsage &daily) {
    out << daily.day << daily.dataUsed << daily.detail;
    return out;
}

QDataStream &operator>>(QDataStream &in, DailyUsage &daily) {
    in >> daily.day >> daily.dataUsed >> daily.detail;
    return in;
}

QDataStream &operator<<(QDataStream &out, const BLUsage &usage) {
    out << usage.name << usage.username << usage.password << usage.start << usage.end << usage.lastUpdate << usage.totalKB << usage.usage;
    return out;
}

QDataStream &operator>>(QDataStream &in, BLUsage &usage) {
    in >> usage.name >> usage.username >> usage.password >> usage.start >> usage.end >> usage.lastUpdate >> usage.totalKB >> usage.usage;
    return in;
}

File: blusage.py
import datetime

from BeautifulSoup import BeautifulSoup

class DailyUsage(object):
    day = ""
    dataUsed = 0
    detail = []

class BLUsage(object):
    name = ""
    username = ""
    password = ""

    last_update = ""

    usage = []
    totalKB = 0
    capKB = 0

    @property
    def error(self):
        return self._error
    _error = ""

    def __init__(self):
        today = datetime.date.today()
        self.start = datetime.date(year=today.year, month=today.month, day=1)
        self.end = self.start + datetime.timedelta(days=29)

    def parse(self, html):
        html = html.lower()

        # attempt to sanitize teh html
        html = html.replace('&nbsp;', '')
        html = html.replace('style=&{head};', '')
        e = [('<head', '</head>'), ('<script', '</script>'), ('<!--', '-->')]
        for elem in e:
            while True:
                begin = html.find(elem[0])
                if begin < 0:
                    break
                end = html.find(elem[1])
                html = html[:begin] + html[end + len(elem[1]):]

        begin = html.find('<body')
        end = html.find('>', begin)
        html = html[:begin + 5] + html[end:]

        soup = BeautifulSoup(unicode(html), convertEntities=BeautifulSoup.HTML_ENTITIES)
        rows  = soup.findAll('table')[14]('tr')

        self.usage = []
        date = None
        timely = []
        dataKB = 0
        self.totalKB = 0
        for row in rows[1:-1]:
            tds = [td.font.contents[0] for td in row('td')[2:]]
            if not date == tds[0]:
                if date:
                    daily = DailyUsage()
                    daily.day = date
                    daily.dataUsed = dataKB
                    daily.detail = timely
                    self.totalKB += dataKB
                    self.usage.append(daily)
                date = tds[0]
                dataKB = 0
                timely = []
            dataKB += int(tds[2])
            timely.append(tds[1:])
        if date:
            daily = DailyUsage()
            daily.day = date
            daily.dataUsed = dataKB
            daily.detail = timely
            self.totalKB += dataKB
            self.usage.append(daily)

        self.last_update = datetime.datetime.now()
        return True

    def smart_bytes(self, data):
        return data

